### PR TITLE
Lesson4 improvement

### PR DESCRIPTION
**Pull request by ElemarJR**

### CODE REVIEW

File: src/Unit-3/lesson4/README.md [hunk 1]
```diff
@@ -108,7 +108,205 @@ SignalR if we are running in web application, for example).
...
+Assume that we have an user busy working on a document. And that process
+can take a while. In the meanwhile, another user came in and changes that
+document. What would be the experience from the point of the user? Well,
+either we let the Last Write Win, or we use Optimistic Concurrency. Either
+way we’re going to have to annoy someone. How about being able to notify
+the user, as soon as the document has updated, that it needs to be refreshed?
+This is a good scenario to use the Changes API. In this excercise you will learn 
+how to do that.
+
+The source code of this exercise is avaiable [here](CollaborativeCategoryEditing).
```
Comment (arekpalinski) [inline src/Unit-3/lesson4/README.md]: typo: avaiable

**Comment by arekpalinski** [comment]
I like the exercise however but I'm not sure if the usage of WinForms was a good choice. You need to handle things like `ToggleEditing`, "setting the Enabled property of NameTextbox, DescriptionTextbox,....  to False" etc which aren't really relevant for this exercise.

Personally I'd prefer a simple console app which would just notify that a document we modified has changed and give a user an option to retrieve it then.

File: src/Unit-3/lesson4/CollaborativeCategoryEditing/EditCategoryForm.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+                ToggleEditing();
+            }
+        }
+
+        private bool DocumentChangedByOtherUser(DocumentChangeNotification change)
+        {
+            if (_savesCount == 0) return true;
+
+            var numberOfServerChanges = change.Etag.Changes - _localEtag.Changes;
```
Comment (arekpalinski) [inline src/Unit-3/lesson4/CollaborativeCategoryEditing/EditCategoryForm.cs:53]: note that on server restart we reset `Changes` so it's potentially possible to get a negative value here

Comment (ElemarJR) [reply src/Unit-3/lesson4/CollaborativeCategoryEditing/EditCategoryForm.cs:53]: If the server restarts the connections to ChangesAPI are restored?

**Comment by ElemarJR** [comment]
@arekpalinski: I agree that the using of a Windows Forms Application adds complexity. Anyway, I think it makes the example more interesting... Don't you think?

File: src/Unit-3/lesson4/CollaborativeCategoryEditing/EditCategoryForm.cs (hunk 2 above)
Comment (arekpalinski) [reply src/Unit-3/lesson4/CollaborativeCategoryEditing/EditCategoryForm.cs:53]: Once we loose the connection we try to renew it then so the client will be notified even after the restart.

**Comment by arekpalinski** [comment]
Yes, it's more interesting - as everything what has the UI. I'm not a UI guy so the console is typically enough for me, not necessary the case for most people :) My feeling was that WinForms code overwhelms the ease of usage of our ChangesAPI.

**Comment by ElemarJR** [comment]
I am not an "UI guy" as well. That layout is something in the limit of what
I can produce. :)
Em qua, 30 de nov de 2016 às 05:43, Arkadiusz Paliński <
notifications@github.com> escreveu:

**Comment by ElemarJR** [comment]
Is there some recommendation of what to do? What should be the approach
here?

Em qua, 30 de nov de 2016 às 05:30, Arkadiusz Paliński <
notifications@github.com> escreveu:

**Comment by arekpalinski** [comment]
Just compare that the etag has changed since we modified it. Use
`_session.Advanced.GetEtagFor(_category)`
after you save it and check if it's different than change.Etag (Etag class
has != operator defined)

**Comment by ElemarJR** [comment]
Would be that enough, including concurrent scenarios?
Em qua, 30 de nov de 2016 às 09:03, Arkadiusz Paliński <
notifications@github.com> escreveu:

> Just compare that the etag has changed since we modified it. Use
> `_session.Advanced.GetEtagFor(_category)`
> after you save it and check if it's different than change.Etag (Etag class
> has != operator defined)
>
> On Wed, Nov 30, 2016 at 11:09 AM Elemar Rodrigues Severo Junior <

**Comment by arekpalinski** [comment]
Not sure what do you mean. Once you successfully updated a doc you are
guaranteed that it was saved with etag you got in the response which is
going to be returned as the result of GetEtagFor call since the doc was
updated within this session.

**Comment by ElemarJR** [comment]
The notification event will be fired before I have change to get the new
Etag. :)

My idea was to have some flag before saving...

Em qua, 30 de nov de 2016 às 09:21, Arkadiusz Paliński <
notifications@github.com> escreveu:

> Not sure what do you mean. Once you successfully updated a doc you are
> guaranteed that it was saved with etag you got in the response which is
> going to be returned as the result of GetEtagFor call since the doc was
> updated within this session.
>
> On Wed, Nov 30, 2016 at 12:08 PM Elemar Rodrigues Severo Junior <

**Comment by arekpalinski** [comment]
You have the new etag right after you call `SaveChanges()`. `GetEtagFor` won't do a remote call then.

However, a race condition it's still possible since notifications are sent right after we commit a transaction on the server. So it may arrive before `SaveChanges` response (not very likely but still). That would require syncing these two pieces of code :/ The worst case scenario without the syncing is that you notify a user about the change but the refresh will retrieve the same doc.

**Comment by ElemarJR** [comment]
I've been getting the change notification before I can get the new ETAG. I
think we should continue to use the current version. What do you think?

Em qua, 30 de nov de 2016 às 12:02, Arkadiusz Paliński <
notifications@github.com> escreveu:

**Comment by arekpalinski** [comment]
We can use the current version but please also add a check comparing
Etag.Restarts numbers to make it working after the server restart.

**Pull request merged**